pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,211 @@
+)
+
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	namespaceLister corev1listers.NamespaceLister
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// NewController initializes the controller and is called by the generated code
+// Registers event handlers to enqueue events
+func NewController(
+	opt reconciler.Options,
+	namespaceInformer corev1informers.NamespaceInformer,
+) *controller.Impl {
+
+	r := &Reconciler{
+		Base:            reconciler.NewBase(opt, controllerAgentName),
+		namespaceLister: namespaceInformer.Lister(),
+	}
+	impl := controller.NewImpl(r, r.Logger, ReconcilerName, reconciler.MustNewStatsReporter(ReconcilerName, r.Logger))
+
+	// TODO: filter label selector: on InjectionEnabledLabels()
+	// TODO: we need to also watch for changes to service accounts, RoleBindings, and Brokers to heal on bad changes.
+
+	r.Logger.Info("Setting up event handlers")
+	namespaceInformer.Informer().AddEventHandler(reconciler.Handler(impl.Enqueue))
+
+	return impl
+}
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Namespace resource
+// with the current status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		r.Logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the namespace resource with this namespace/name
+	original, err := r.namespaceLister.Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("namespace key in work queue no longer exists", zap.Any("key", key))
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if original.Labels[resources.InjectionLabelKey] != resources.InjectionEnabledLabelValue {
+		logging.FromContext(ctx).Debug("Not reconciling Namespace")
+		// TODO: this does not handle cleanup of unwanted brokers in namespace.
+		return nil
+	}
+
+	// Don't modify the informers copy
+	ns := original.DeepCopy()
+
+	// Reconcile this copy of the Namespace and then write back any status updates regardless of
+	// whether the reconcile error out.
+	err = r.reconcile(ctx, ns)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error reconciling Namespace", zap.Error(err), zap.Any("key", key))
+	} else {
+		logging.FromContext(ctx).Debug("Namespace reconciled", zap.Any("key", key))
+	}
+
+	// Requeue if the resource is not ready:
+	return err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	if ns.DeletionTimestamp != nil {
+		return nil
+	}
+	sa, err := r.reconcileBrokerFilterServiceAccount(ctx, ns)
+	if err != nil {
+		logging.FromContext(ctx).Error("Unable to reconcile the Broker Filter Service Account for the namespace", zap.Error(err))
```

Comment:
Would it be useful to emit events when things fail?

Reply:
it would, we should do it as a followup.

Reply:
Added to the issue https://github.com/knative/eventing/issues/1080